ndt7/server/upload/upload.go

```diff
@@ -0,0 +1,206 @@
+	// some weird cases it MAY happen that the cache pruning mechanism will
+	// remove the fp BEFORE we can steal it. In case we cannot get a file
+	// we just abort the test, as this should not happen (TM).
+	if fp == nil {
+		err := errors.New("cannot get file bound to websocket conn")
+		return nil, err
+	}
+	return fp, nil
+}
+
+// gatherAndSaveTCPInfo gathers TCP info measurements from |sockfp| and stores
+// them into the |measurement| object as well as into the |resultfp| file.
+// It returns an error on failure and nil in case of success.
+func gatherAndSaveTCPInfo(measurement *model.Measurement, sockfp *os.File, resultfp *results.File) error {
+	metrics, err := tcpinfox.GetTCPInfo(sockfp)
+	if err == nil {
+		measurement.TCPInfo = &metrics
+	}
+
+	if err := resultfp.WriteMeasurement(*measurement, "server"); err != nil {
+		logging.Logger.WithError(err).Warn("Cannot save measurement on disk")
+		return err
+	}
+	return nil
+}
+
+// measuringLoop is the loop that runs the measurements in a goroutine.
+// This function exits when
+//     (1) a fatal error occurs or
+//     (2) the maximum elapsed time for the upload test expires.
+// The rest of the upload code is supposed to stop receiving when this function
+// signals that we're done by closing the channel.
+func measuringLoop(ctx context.Context, request *http.Request,
+	conn *websocket.Conn, dataDir string, dst chan model.Measurement) {
+	logging.Logger.Debug("Starting measurement loop")
+	defer logging.Logger.Debug("Stopping measurement loop")
+	defer close(dst)
+
+	resultfp, err := results.OpenFor(request, conn, dataDir, "upload")
+	if err != nil {
+		return // error printed already
+	}
+
+	defer iox.ErrorLoggingCloser(resultfp).Close()
+
+	sockfp, err := getConnFile(conn)
+	if err != nil {
+		return // error printed already
+	}
+
+	defer iox.ErrorLoggingCloser(sockfp).Close()
+	t0 := time.Now()
+	ticker := time.NewTicker(spec.MinMeasurementInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case now := <-ticker.C:
+			elapsed := now.Sub(t0)
+			if elapsed > defaultDuration {
+				logging.Logger.Debug("Upload has run for long enough.")
+				return
+			}
+
+			measurement := model.Measurement{
+				Elapsed: elapsed.Seconds(),
+			}
+
+			err = gatherAndSaveTCPInfo(&measurement, sockfp, resultfp)
+			if err != nil {
+				return // error printed already
+			}
+
+			dst <- measurement
+		}
+	}
+}
+
+// startMeasuring runs the measurement loop. This runs in a separate goroutine
+// and emits Measurement events on the returned channel.
+func startMeasuring(ctx context.Context, request *http.Request, conn *websocket.Conn, dataDir string) chan model.Measurement {
+	dst := make(chan model.Measurement)
+	go measuringLoop(ctx, request, conn, dataDir, dst)
+	return dst
+}
+
+// ServeHTTP handles the upload subtest.
+func (ul Handler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	// TODO(evfirerob): factor out this stuff as it's the same for both UL/DL
```

Comment:
(We don't know for sure until the code in here has taken its final shape. Let's remember this.)